connect-inject/health_check_controller.go

```diff
@@ -0,0 +1,260 @@
+	// to a watched or listed resource
+	for c.processNextItem() {
+		c.Log.Debug("Controller.runWorker: processing next item")
+	}
+	c.Log.Debug("Controller.runWorker: completed")
+}
+
+// processNextItem retrieves each Queued item and takes the
+// necessary Handle action based off of if the item was created, updated or deleted
+func (c *HealthCheckController) processNextItem() bool {
+	c.Log.Debug("Controller.processNextItem: start")
+
+	// fetch the next item (blocking) from the Queue to process or
+	// if a shutdown is requested then return out to stop
+	key, quit := c.Queue.Get()
+	if quit {
+		c.Log.Error("controller.processNextItem: shutting down")
+		return false
+	}
+	// Key format is as follows :  CREATE/namespace/name, DELETE/namespace/name, UPDATE/namespace/name
+	// also keep track if this is an create
+	create := true
+	formattedKey := strings.Split(key.(string), "/")
+	if formattedKey[0] != "ADD" {
+		create = false
+	}
+	keyRaw := strings.Join(formattedKey[1:], "/")
+
+	// item will contain the complex object for the resource and
+	// exists is a bool that'll indicate whether or not the
+	// resource was created (true) or deleted (false)
+	//
+	// if there is an error in getting the key from the index
+	// then we want to retry this particular Queue key a certain
+	// number of times (c.MaxRetries) before we forget the Queue key
+	// and throw an error
+	item, exists, err := c.Informer.GetIndexer().GetByKey(keyRaw)
+	if err != nil {
+		if c.Queue.NumRequeues(key) < c.MaxRetries {
+			c.Log.Error("controller.processNextItem: Failed processing item with key %s with error %v, retrying", key, err)
+			c.Queue.AddRateLimited(key)
+		} else {
+			c.Log.Error("controller.processNextItem: Failed processing item with key %s with error %v, no more retries", key, err)
+			c.Queue.Forget(key)
+			utilruntime.HandleError(err)
+		}
+	}
+
+	// if the object does exist that indicates that the object
+	// was created or updated so run the ObjectCreated/ObjectUpdated method
+	// dequeue the key to indicate success, requeue it on failure
+	if exists {
+		if create == true {
+			// This is a Pod Create
+			c.Log.Debug("controller.processNextItem: object create detected: %s", keyRaw)
+			err = c.Handle.ObjectCreated(item)
+			if err == nil {
+				c.Log.Debug("controller.processNextItem: object update as part of ObjectCreate: %s", keyRaw)
+				err = c.Handle.ObjectUpdated(item)
+			}
+		} else {
+			// This is a Pod Status Update
+			c.Log.Debug("controller.processNextItem: object update detected: %s", keyRaw)
+			err = c.Handle.ObjectUpdated(item)
+		}
+		if err == nil {
+			// Indicates success
+			c.Queue.Forget(key)
+		} else if c.Queue.NumRequeues(key) < c.MaxRetries {
+			c.Log.Error("unable to process request, retrying")
+			c.Queue.AddRateLimited(key)
+		}
+	}
+	if err == nil {
+		// marking Done removes the key from the queue entirely
+		c.Queue.Done(key)
+	}
+	// keep the worker loop running by returning true
+	return true
+}
```

Comment:
It looks like a lot of this code was copied from `helper/controller.go`. Is there a way to share some of the code from there?